Adding Lambda Tutorial for OCI Images (#1420)
aws-controllers-k8s/community: docs/content/docs/tutorials/lambda-oci-example.md

### Prerequisites

This guide assumes that you have:

- Created an EKS cluster with Kubernetes version 1.16 or higher.
- Have access to Amazon ECR
- AWS IAM permissions to create roles and attach policies to roles.
- Installed the following tools on the client machine used to access your Kubernetes cluster:
  - [AWS CLI](https://docs.aws.amazon.com/cli/latest/userguide/install-cliv1.html) - A command line tool for interacting with AWS services.
  - [kubectl](https://docs.aws.amazon.com/eks/latest/userguide/install-kubectl.html) - A command line tool for working with Kubernetes clusters.
  - [eksctl](https://docs.aws.amazon.com/eks/latest/userguide/eksctl.html) - A command line tool for working with EKS clusters.
  - [Helm 3.8+](https://helm.sh/docs/intro/install/) - A tool for installing and managing Kubernetes applications.
  - [Docker](https://docs.docker.com/engine/install/) - A tool to build, share, and run containers.

### Install the ACK service controller for Lambda

Log into the Helm registry that stores the ACK charts:
```bash
aws ecr-public get-login-password --region us-east-1 | helm registry login --username AWS --password-stdin public.ecr.aws
```

> Deploy the ACK service controller for Amazon Lambda using the [lambda-chart Helm chart](https://gallery.ecr.aws/aws-controllers-k8s/lambda-chart). Resources should be created in the `us-east-1` region:

Reviewer: ```suggestion
Deploy the ACK service controller for Amazon Lambda using the [lambda-chart Helm chart](https://gallery.ecr.aws/aws-controllers-k8s/lambda-chart). This example creates resources in the `us-east-1` region, but you can use any other region supported in AWS.
```

```

For a full list of available values to the Helm chart, please [review the values.yaml file](https://github.com/aws-controllers-k8s/lambda-controller/blob/main/helm/values.yaml).

### Configure IAM permissions

Once the service controller is deployed [configure the IAM permissions](https://aws-controllers-k8s.github.io/community/docs/user-docs/irsa/) for the
controller to invoke the Lambda API. For full details, please review the AWS Controllers for Kubernetes documentation
for [how to configure the IAM permissions](https://aws-controllers-k8s.github.io/community/docs/user-docs/irsa/). If you follow the examples in the documentation, use the
value of `lambda` for `SERVICE`.

## Create Lambda function handler
The Lambda [function handler](https://docs.aws.amazon.com/lambda/latest/dg/nodejs-handler.html) is the method in your function code that processes events. When your function is invoked, Lambda runs the handler method.

```bash
cat <<EOF > app.js
exports.handler = async (event) => {
    const response = {
        statusCode: 200,

>         body: JSON.stringify('Hello from Lambda!'),

Reviewer: ```suggestion
        body: JSON.stringify('Hello from Lambda!')
```

(My JavaScript knowledge may be fading, but IIRC you're not supposed to have extra commas in object definitions).

    };
    return response;
};
EOF
```

## Create and Build a Docker Image

> Create a Dockerfile to build an image from. 

Reviewer: ```suggestion
Create a Dockerfile that will be used to build the image that will be used for our Lambda function:
```

```bash
cat <<EOF > Dockerfile
FROM public.ecr.aws/lambda/nodejs:14

COPY app.js package.json ./

RUN npm install

CMD [ "app.handler" ]
EOF
```

> Install dependencies and build a Docker image called hello-world.

Reviewer: ```suggestion
Build the Docker image in your local environment. You will need to install dependencies using `npm`:
```

```shell
npm init -y
docker build -t hello-world .
```
## Publish the Docker image to ECR

> Publish the Docker image to an ECR repository. It's a requirement for container images to be published to the ECR repository to run Lambda OCI image functions.

Reviewer: Double checking if using the ECR repository is required for this or if it's just recommended for the tutorial.

Author: Yes, its required

```shell
export AWS_ACCOUNT_ID=$(aws sts get-caller-identity --query "Account" --output text)
export AWS_REGION=us-west-2

> aws ecr get-login-password --region us-west-2 | docker login --username AWS --password-stdin ${AWS_ACCOUNT_ID}.dkr.ecr.${AWS_REGION}.amazonaws.com

Reviewer: Two things:

1. Is there a reason why the tutorial switches to `us-west-2` here when we talk about setting up the controller for `us-east-1`?
2. There is mixed use of shell variables and hard coded strings here, e.g. `us-west-2` is spelled out in the beginning.
3. I recommend wrapping the strings that have variables, i.e. `"${AWS_ACCOUNT_ID}.dkr.ecr.${AWS_REGION}.amazonaws.com"`

Author: Good catch. Will redefine.

aws ecr create-repository --repository-name hello-world --image-scanning-configuration scanOnPush=true --image-tag-mutability MUTABLE
docker tag  hello-world:latest ${AWS_ACCOUNT_ID}.dkr.ecr.${AWS_REGION}.amazonaws.com/hello-world:latest

> docker push ${AWS_ACCOUNT_ID}.dkr.ecr.${AWS_REGION}.amazonaws.com/hello-world:latest

Reviewer: Wrap strings that have variables.

```

## Deploy the Lambda OCI function using the ACK Lambda controller

> Execute the following commands to create a manifest containing the Lambda OCI function and submit this manifest to the EKS cluster using kubectl.

Reviewer: ```suggestion
The following example creates a manifest that contains the Lambda OCI function. It then uses `kubectl` to create the resource in Kubernetes:
```

```shell
export AWS_ACCOUNT_ID=$(aws sts get-caller-identity --query "Account" --output text)
export IMAGE_URI="${AWS_ACCOUNT_ID}.dkr.ecr.${AWS_REGION}.amazonaws.com/hello-world:latest "
export FUNCTION_NAME="lambda-oci-ack"
export LAMBDA_ROLE="arn:aws:iam::${AWS_ACCOUNT_ID}:role/lambda_basic_execution"

read -r -d '' LAMBDA_MANIFEST <<EOF
apiVersion: lambda.services.k8s.aws/v1alpha1
kind: Function
metadata:
 name: $FUNCTION_NAME
 annotations:
   services.k8s.aws/region: $AWS_REGION
spec:
 name: $FUNCTION_NAME
 packageType: Image
 code:
     imageURI: $IMAGE_URI
 role: $LAMBDA_ROLE
 description: function created by ACK lambda-controller e2e tests
EOF

echo "${LAMBDA_MANIFEST}" > function.yaml

kubectl create -f function.yaml
```
You should get a confirmation that the function was created successfully.

```
function.lambda.services.k8s.aws/lambda-oci-ack created
```
To get details about the Lambda function, run the following.

```bash

> kubectl describe function/${FUNCTION_NAME}

Reviewer: ```suggestion
kubectl describe "function/${FUNCTION_NAME}"
```

```

## Invoke the Lambda OCI Function

> After you have verified that the Lambda OCI function is deployed correctly, you can now invoke your function through the [AWS CLI](https://docs.aws.amazon.com/cli/latest/reference/lambda/index.html).

Reviewer: ```suggestion
After you have verified that the Lambda OCI function is deployed correctly, you can invoke the function through the [AWS CLI](https://docs.aws.amazon.com/cli/latest/reference/lambda/index.html).
```

```bash
aws lambda invoke --function-name ${FUNCTION_NAME} response.json
cat response.json
```

You will get the output as below:
```
{"statusCode":200,"body":"\"Hello from Lambda!\""} 
```

## Next steps

The ACK service controller for Amazon Lambda is based on the [Amazon Lambda API](https://docs.aws.amazon.com/lambda/latest/dg/API_Reference.html).

Refer to [API Reference](https://aws-controllers-k8s.github.io/community/reference/) for *Lambda* to find
all the supported Kubernetes custom resources and fields.

### Cleanup

> You can delete your Lambda OCI function using the `kubectl delete` command.

Reviewer: ```suggestion
You can delete your Lambda OCI function using the `kubectl delete` command:
```

  - [AWS CLI](https://docs.aws.amazon.com/cli/latest/userguide/install-cliv1.html) - A command line tool for interacting with AWS services.
  - [kubectl](https://docs.aws.amazon.com/eks/latest/userguide/install-kubectl.html) - A command line tool for working with Kubernetes clusters.
  - [eksctl](https://docs.aws.amazon.com/eks/latest/userguide/eksctl.html) - A command line tool for working with EKS clusters.
  - [Helm 3.8+](https://helm.sh/docs/intro/install/) - A tool for installing and managing Kubernetes applications.
  - [Docker](https://docs.docker.com/engine/install/) - A tool to build, share, and run containers.

### Install the ACK service controller for Lambda

Log into the Helm registry that stores the ACK charts:

```bash
aws ecr-public get-login-password --region us-west-2 | helm registry login --username AWS --password-stdin public.ecr.aws
```

Deploy the ACK service controller for Amazon Lambda using the [lambda-chart Helm chart](https://gallery.ecr.aws/aws-controllers-k8s/lambda-chart). This example creates resources in the `us-west-2` region, but you can use any other region supported in AWS.

```bash

> export RELEASE_VERSION=`curl -sL https://api.github.com/repos/aws-controllers-k8s/$SERVICE-controller/releases/latest | grep '"tag_name":' | cut -d'"' -f4`

Reviewer: ```suggestion
SERVICE=lambda
RELEASE_VERSION=$(curl -sL "https://api.github.com/repos/aws-controllers-k8s/${SERVICE}-controller/releases/latest" | grep '"tag_name":' | cut -d'"' -f4)
```

export IMAGE_URI="${AWS_ACCOUNT_ID}.dkr.ecr.${AWS_REGION}.amazonaws.com/hello-world:latest "
export FUNCTION_NAME="lambda-oci-ack"
export LAMBDA_ROLE="arn:aws:iam::${AWS_ACCOUNT_ID}:role/lambda_basic_execution"

read -r -d '' LAMBDA_MANIFEST <<EOF
apiVersion: lambda.services.k8s.aws/v1alpha1
kind: Function
metadata:
 name: $FUNCTION_NAME
 annotations:
   services.k8s.aws/region: $AWS_REGION
spec:
 name: $FUNCTION_NAME
 packageType: Image
 code:
     imageURI: $IMAGE_URI
 role: $LAMBDA_ROLE
 description: function created by ACK lambda-controller e2e tests
EOF

echo "${LAMBDA_MANIFEST}" > function.yaml

kubectl create -f function.yaml
```
You should get a confirmation that the function was created successfully.

```
function.lambda.services.k8s.aws/lambda-oci-ack created
```
To get details about the Lambda function, run the following.

```bash
kubectl describe "function/${FUNCTION_NAME}"
```

## Invoke the Lambda OCI Function
After you have verified that the Lambda OCI function is deployed correctly, you can invoke the function through the [AWS CLI](https://docs.aws.amazon.com/cli/latest/reference/lambda/index.html).

```bash

> aws lambda invoke --function-name ${FUNCTION_NAME} response.json

Reviewer 2: tiny nit: you can also use `aws lambda invoke --function-name --region us-west-2 demo-function-list-items /dev/stdout | jq`
